generator/feature_sorter.cpp

```diff
@@ -174,71 +179,82 @@ class FeaturesCollector2 : public FeaturesCollector
         {
           // Simplify and serialize triangles.
-          bool const good = isCoast || scales::IsGoodOutlineForLevel(level, points);
+          Polygons simplified;
 
-          // At this point we don't need last point equal to first.
-          CHECK_GREATER(points.size(), 0, ());
-          points.pop_back();
+          bool const isGood = isCoast || scales::IsGoodOutlineForLevel(level, points);
+          if (isGood)
+          {
+            // We don't need the last point equal to the first.
+            CHECK_GREATER(points.size(), 3, ());
+            points.pop_back();
 
-          Polygons const & polys = fb.GetGeometry();
-          if (polys.size() == 1 && good && holder.TryToMakeStrip(points))
-            continue;
+            if (polys.size() == 1 && holder.TryToMakeStrip(points))
+            {
+              /// @todo I suspect we already can break from the loop here, because
+              /// successful TryToMakeStrip makes NeedProcessTriangles() == false.
+              continue;
+            }
 
-          Polygons simplified;
-          if (good)
-          {
-            simplified.push_back({});
-            simplified.back().swap(points);
+            simplified.push_back(std::move(points));
           }
           else
-            LOG(LDEBUG, ("Area: too small or degenerate 1st polygon of", polys.size(), ", points count", points.size(), "at scale", i, DebugPrint(fb)));
+          {
+            /// @todo Return (or continue) immediately if area is not _good_?
+            LOG(LDEBUG, ("Area: too small or degenerate 1st (outer) polygon of", polys.size(),
+                         ", points count", points.size(), "at scale", i, DebugPrint(fb)));
+          }
 
           auto iH = polys.begin();
           for (++iH; iH != polys.end(); ++iH)
           {
-            simplified.push_back({});
+            points.clear();
+            SimplifyPoints(level, isCoast, rect, *iH, points);
 
-            SimplifyPoints(level, isCoast, rect, *iH, simplified.back());
-
-            if (scales::IsGoodOutlineForLevel(level, simplified.back()))
+            if (scales::IsGoodOutlineForLevel(level, points))
             {
-              // At this point we don't need last point equal to first.
-              CHECK_GREATER(simplified.back().size(), 0, ());
-              simplified.back().pop_back();
+              // We don't need the last point equal to the first.
+              CHECK_GREATER(points.size(), 3, ());
+              points.pop_back();
+              simplified.push_back(std::move(points));
             }
             else
             {
-              // Remove small or degenerate polygon.
-              simplified.pop_back();
-              LOG(LDEBUG, ("Area: too small or degenerate 2nd+ polygon of", polys.size(), ", points count", simplified.back().size(), "at scale", i, DebugPrint(fb)));
+              LOG(LDEBUG, ("Area: too small or degenerate 2nd+ (inner) polygon of", polys.size(),
```

Comment:
Yes, that is the point! Outer polygon is 1st, others are inners and I can't remember why it was done like this. 

We now split multiple outers into several FeatureBuilders. And it is also controversial logic :) because we have several park features (for example) with names where it is one logical feature.